### ISSUE TITLE
HTML syntax on prettier whitespace sensitivty

### ISSUE THREAD

**Issue by jackysee** [vim/vim#4150]
The html below would cause syntax for the `<style>` section not working.
```
<!DOCTYPE html>
<html lang="en">
<body>
    <span class="dolorum atque aspernatur"><a href="abc"
        >Est molestiae sunt facilis qui rem.</a
    ></span>
    <style>
    div{ font-weight:bold }
    </style>
</body>
</html>
```
The screen capture:
![image](https://user-images.githubusercontent.com/30516/54832617-77ad7400-4cf7-11e9-9dd2-3c5302a1c528.png)

The syntax would be correct if the closing `</a>` has no whitespace

![image](https://user-images.githubusercontent.com/30516/54833070-539e6280-4cf8-11e9-84eb-b486667091e9.png)

This kind of formatting of not losing whitespace is now a default in prettier ( https://prettier.io/blog/2018/11/07/1.15.0.html#whitespace-sensitive-formatting ). 

In Vue, the official cli program would generate component where the `<style>` would be placed at the bottom part.

**Comment by brammool** [comment]
> The html below would cause syntax for the `<style>` section not working.
> ```
> <!DOCTYPE html>
> <html lang="en">
> <body>
>     <span class="dolorum atque aspernatur"><a href="abc"
>         >Est molestiae sunt facilis qui rem.</a
>     ></span>
>     <style>
>     div{ font-weight:bold }
>     </style>
> </body>
> </html>
> ```
> The screen capture:
> ![image](https://user-images.githubusercontent.com/30516/54832617-77ad7400-4cf7-11e9-9dd2-3c5302a1c528.png)
> 
> The syntax would be correct if the closing `</a>` has no whitespace
> 
> ![image](https://user-images.githubusercontent.com/30516/54833070-539e6280-4cf8-11e9-84eb-b486667091e9.png)
> 
> 
> This kind of formatting of not losing whitespace is now a default in prettier ( https://prettier.io/blog/2018/11/07/1.15.0.html#whitespace-sensitive-formatting ). 

The example has a line break, but the same happens when the closing tag
is </a >, while it looks OK with </a>.

Jorge, can you fix this?

-- 
Contrary to popular belief, Unix is user friendly.
It just happens to be selective about who it makes friends with.
                                               -- Dave Parnas

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by jorgesumle** [comment]
This issue applies to every region inside `if !exists("html_no_rendering")`. In this case, on the [line 194](https://github.com/vim/vim/blob/master/runtime/syntax/html.vim#L194). When you add a space before the `>`, `me=e-4` is wrong, and it would need to be `me=e-5` for it to work. I don't know if it's possible to do something like `me=e-[end-match-length]`, so that there are no highlighting problems, I didn't find such an option in the documentation. Also I didn't code that part and I don't know if there is a better way to highlight.

> Jorge, can you fix this?

As you've seen, I tried without sucess. I won't work more on this because I lack the time, but I hope that this comment will help and I'm happy to review any patch which aims to solve this issue.

**Comment by tom-m** [comment]
I took the [line 194](https://github.com/vim/vim/blob/master/runtime/syntax/html.vim#L194), where the `htmlLink` region is defined:
```vim
syn region htmlLink start="<a\>\_[^>]*\<href\>" end="</a>"me=e-4 contains=@Spell,htmlTag,htmlEndTag,htmlSpecialChar,htmlPreProc,htmlComment,htmlLeadingSpace,javaScript,@htmlPreproc
```
And changed `end="</a>"me=e-4` to `end="</a\_s*>"me=s-1`. After this the OP's html code highlighting worked great.

As my knowledge of html is very poor, I'm not able to do proper testing. So I cannot tell if this change breaks anything. What do you think, @jorgesumle or anybody? Might this be the way to go?

Tom

**Comment by jorgesumle** [comment]
> What do you think, @jorgesumle or anybody? Might this be the way to go?

Sorry, I didn't read the documentation carefully enough. Sure, it does exactly what we want and it's more flexible than the previous code where the tag length was needed to proper highlight and just one space could mess up everything. Thanks a lot for the fix @tom-m!

Could you please send a pull request [to the repository](https://github.com/vim/vim/blob/master/runtime/syntax/html.vim#L5) (or email me using [git request-pull](https://git-scm.com/docs/git-request-pull) or a patch, I don't really mind)? When it's ready, I'll send the patch to Bram so that we can solve this issue.

**Comment by jorgesumle** [comment]
After more than a week, I still have no answer from you, @tom-m. If you don't answer in two days, I'll assume you're not working on the patch and I'll work on it to fix this issue.

**Comment by tom-m** [comment]
@jorgesumle

Sorry for answering late, Jorge. I was not watching this thread too closely. Anyway, as I stated before, I'm no HTML guru. I cannot tell for sure where does HTML allow whitespace. So I didn't feel to be in a position to propose changes to the syntax file myself. Thank you for coming up with proper fix for this issue.

Tom

**Comment by jorgesumle** [comment]
Don't worry, @tom-m, thanks for your help!

> I cannot tell for sure where does HTML allow whitespace.

[According to the W3C HTML Recommendation](https://www.w3.org/TR/html52/syntax.html#end-tags): "After the [end] tag name, there may be one or more space characters".

**Comment by chrisbra** [comment]
closed by #4219
